cub/agent/agent_batch_memcpy.cuh

```diff
@@ -0,0 +1,967 @@
+
+  LoadVectorAndFunnelShiftR(aligned_ptr, bit_shift, data_out);
+}
+
+/**
+ * @brief Cooperatively copies \p num_bytes from \p src to \p dest using vectorized stores of type \p VectorT for
+ * addresses within [dest, dest + num_bytes) that are aligned to \p VectorT. A byte-wise copy is used for byte-ranges
+ * that are not aligned to \p VectorT.
+ *
+ * @tparam VectorT The vector type used for vectorized stores (i.e., one of uint4, uint2, uint32_t)
+ * @tparam ByteOffsetT Type sufficiently large to
+ * @param thread_rank The thread rank within the group that cooperates to copy the data must be within [0, `group_size`)
+ * @param group_size The number of threads cooperaing to copy the data; all threads within [0, `group_size`) must invoke
+ * this method with the same arguments
+ * @param dest Pointer to the memory location to copy to
+ * @param num_bytes Number of bytes to copy
+ * @param src Pointer to the memory location to copy from
+ */
+template <typename VectorT, typename ByteOffsetT>
+__device__ __forceinline__ void
+VectorizedCopy(int32_t thread_rank, int32_t group_size, void *dest, ByteOffsetT num_bytes, const void *src)
+{
+  char *out_ptr                      = reinterpret_cast<char *>(dest);
+  const char *in_ptr                 = reinterpret_cast<const char *>(src);
+  constexpr size_t out_datatype_size = sizeof(VectorT);
+  constexpr size_t in_datatype_size  = sizeof(uint32_t);
+
+  // Number of bytes between out_ptr and the first OutT-aligned address at or after out_ptr
+  auto const alignment_offset = reinterpret_cast<std::uintptr_t>(out_ptr) % out_datatype_size;
+
+  // The first OutT-aligned address before (or at) out_ptr
+  VectorT *out_chars_aligned = reinterpret_cast<VectorT *>(out_ptr - alignment_offset);
+
+  // Both `out_start_aligned` and `out_end_aligned` are indices into `out_ptr`.
+  // `out_start_aligned` is the first OutT-aligned memory location after `out_ptr + 4`.
+  // `out_end_aligned` is the last OutT-aligned memory location before `out_end - 4`, where out_end corresponds to one
+  // past the last byte to be copied. Bytes between `[out_start_aligned, out_end_aligned)` will be copied using
+  // uint4. `out_ptr + 4` and `out_end - 4` are used instead of `out_ptr` and `out_end` to avoid `LoadVector`
+  // reading beyond data boundaries.
+  int32_t out_start_aligned = (in_datatype_size + alignment_offset + out_datatype_size - 1) / out_datatype_size *
+                                out_datatype_size -
+                              alignment_offset;
+  int32_t out_end_aligned = (num_bytes - in_datatype_size + alignment_offset) / out_datatype_size * out_datatype_size -
+                            alignment_offset;
+
+  // Perform vectorized copies by loading 4B at the time to populate a `VectorT` and write that out to global memory
+  for (int64_t ichar = out_start_aligned + thread_rank * out_datatype_size; ichar < out_end_aligned;
+       ichar += group_size * out_datatype_size)
+  {
+    VectorT data_in;
```

Comment:
I am unsure about the magic cuda does in the background, but could we hoist `data_in` out of the loop?

Reply:
No strong opinion; but let me give my motivation and we can re-assess what's the best solution 🙂  

I think currently it's obvious that the scope of the variable is limited to the loop and that each loop iteration's `data_in` is independent from the previous iteration. I would assume that this makes it easier for the reader to understand the logic and for the compiler to vectorise across loop iterations. Wdyt?

Reply:
I am fine with both. 

Truth be told, I am still heavily in the `T could be everything imaginable` mode from pure C++ rather than `T is a primitive type` that we deal with here. 

So I assume a) the compiler should take care of it. and b) even if not the drawback would be nil. 

Also you are right it clearly shows that data is independent.